src/error/Error.h

```diff
@@ -11,29 +11,53 @@
 
 #include "error/Detail.h"
 #include "error/forward.h"
+#include "mem/PoolingAllocator.h"
 
 #include <iosfwd>
+#include <vector>
+
+/// zero or more details of a single error
+using ErrorDetails = std::vector<ErrorDetailPointer, PoolingAllocator<ErrorDetailPointer> >;
```

Comment:
FWIW, I have tried to implement this using std::list, but it only resulted in more memory allocations/deallocations (because a list has to allocate and deallocate every list node while a vector allocates and deallocates all nodes at once).

Lists can be spliced with fewer memory operations, but the current branch code no longer needs to splice, and the earlier std::forward_list code that was still splicing was not really outperforming an std::vector implementation either (both had about the same number of memory operations for a few errors and std::vector was often winning when handling more details). The relevant use cases just do not benefit from using lists, even though std::vector has to occasionally reallocate.

STL lists may be used where container elements are expensive or impossible to copy; ErrorDetail::Pointer is cheap to copy. STL lists may also be more efficient when we need to delete or add nodes in the middle of a container, but we do not need to do that when storing error details.